From c3f99f84a55641bbb968c468d0727bcecb274312 Mon Sep 17 00:00:00 2001
From: Naveen Mamindlapalli <naveenm@marvell.com>
Date: Mon, 17 Aug 2020 12:07:19 +0530
Subject: [PATCH 4/5] octeontx2-bphy-netdev: Set minimum length of Tx packets
 to 64 bytes

commit a0e9c192c3c95b7c6e0442c9741902ffcf25bc19 from
git@git.assembla.com:cavium/WindRiver.linux.git

This patch sets the minimum length of CPRI Tx packets to 64 bytes and
zero pad the buffer in case of short packets to avoid sending junk data
since there is no support for zero padding in the CPRI HW. This patch
fixes issue of transmitting out short packets such as ARP packets.

Change-Id: I122775a0f08a36cffd3391ba3ea29720c2e01351
Signed-off-by: Naveen Mamindlapalli <naveenm@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/34091
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
---
 drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c
index cc04ac1a0ebd..a56397916950 100644
--- a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c
+++ b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c
@@ -397,8 +397,11 @@ static netdev_tx_t otx2_cpri_eth_start_xmit(struct sk_buff *skb,
 	wqe->mhab_id = priv->cpri_num;
 	wqe->lane_id = priv->lmac_id;
 	buf_ptr += OTX2_BPHY_CPRI_WQE_SIZE;
+	/* zero pad for short pkts, since there is no HW support */
+	if (skb->len < 64)
+		memset(buf_ptr, 0, 64);
 	memcpy(buf_ptr, skb->data, skb->len);
-	wqe->pkt_length = skb->len;
+	wqe->pkt_length = skb->len > 64 ? skb->len : 64;
 
 	/* ensure the memory is updated before ringing doorbell */
 	dma_wmb();
-- 
2.17.1

